runners/direct-java/src/test/java/org/apache/beam/runners/direct/ExecutorServiceParallelExecutorTest.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.direct;
+
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.emptySet;
+
+import com.google.common.collect.LinkedListMultimap;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.testing.ThreadLeakTracker;
+import org.joda.time.Instant;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestName;
+import org.junit.rules.TestRule;
+
+/**
+ * Validates some basic behavior for the ExecutorServiceParallelExecutor.
+ */
+public class ExecutorServiceParallelExecutorTest {
+
+  @Rule
+  public final TestRule threadTracker = new ThreadLeakTracker();
+
+  @Rule
+  public final TestName testName = new TestName();
+
+  @Test
+  public void ensureMetricsThreadDoesntLeak() {
```

Comment:
FYI, this test failed on an unrelated precommit: https://github.com/apache/beam/pull/5218 https://scans.gradle.com/s/rzxuq4ibgr32a/tests/jqhvlvf72f7pg-ipde5etqqejoa. Is the fix in https://github.com/apache/beam/pull/5096 ?

Reply:
@swegner Thanks for pointing out. The test failure shows that the thread created to fake a metrics call in ExecutorServiceParallelExecutorTest#ensureMetricsThreadDoesntLeak sometimes leak. This can happen if this thread is not stopped before the timeout of 10s after the call of shutdown() on its metricsExecutorService. The call of shutdown() is done inside the test itself via the call of stop() on ExecutorServiceParallelExecutor so I don't think #5096 could fix it.  @rmannibucau any comment on #5096?